File: 735-asteroid-collision/asteroid-collision.java
class Solution {
public static int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> collision = new Stack<>();
        for(int i : asteroids) {
            if(collision.isEmpty()) {
                collision.push(i);
            } else {
                int top = collision.peek();
                boolean didExplode = false;

                while(top != -1 && !didExplode && (isPositive(top) && isNegative(i)))  {
                    if(didExplode) break;
                    if(Math.abs(top) > Math.abs(i)) {
                        didExplode = true;
                        break;
                    } else if(Math.abs(top) < Math.abs(i)) {
                        collision.pop();
                    } else {
                        collision.pop();
                        didExplode = true;
                    }
                    top = collision.isEmpty() ? -1 : collision.peek();
                }

                if(!didExplode) {
                    collision.push(i);
                }
            }
        }
        int[] res = new int[collision.size()];
        for(int i = res.length - 1; i >= 0; i--) {
            res[i] = collision.pop();
        }
        return res;
    }

    public static boolean isNegative(int asteroid) {
        return asteroid < 0;
    }

    public static boolean isPositive(int asteroid) {
        return asteroid > 0;
    }
}

File: 735-asteroid-collision/asteroid-collision.py
class Solution(object):
    def asteroidCollision(self, asteroids):
        stack = []
        for i in asteroids:
            if len(stack) == 0:
                stack.append(i)
            else:
                shouldInsert = True
                while len(stack) > 0 and (self.isPositive(stack[-1]) and not self.isPositive(i)):
                    if abs(stack[-1]) < abs(i):
                        stack.pop()
                    elif abs(stack[-1]) == abs(i):
                        shouldInsert = False
                        stack.pop()
                        break
                    else:
                        shouldInsert = False
                        break
                    
                if shouldInsert:
                    stack.append(i)
                
        return stack

    def isPositive(self,num):
        return True if (num * 1) > 0 else False
        